### ISSUE TITLE
TS Error specifically with `lean` result

### ISSUE THREAD

**Issue by iammola** [Automattic/mongoose#11904]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

6.3.5

### Node.js version

18.2.0

### MongoDB server version

4.5.0

### Description

Firstly, Thanks for the amount of effort you put into this package.

The way the new lean type is computed is very frustrating to use. With the object keys it omits from the type and repeated computed names, it makes it hard to understand what the actual result type will be by just hovering over the variable.

I upgraded from `v6.3.1` to `v6.3.5` and the changes broke the lean result type in my application. The `ObjectId` type I use is defined with the one in `bson` as well.

```ts
import { ObjectId } from "bson";
import { Schema } from "mongoose";

type ObjectId = ObjectId & Schema.Types.ObjectId;
```

I do it this way to take advantage of the `bson` type definition for an `ObjectId`. Like with the `.equals` method to test the equality of an `ObjectIdLike` variable with the ObjectId. TS reported the property doesn't exist in the lean return type, but I tested and it did.

 I edited the definition of your types for the LeanResult to something more straightforward, the error was resolved and it was easier to glance at the type when I hovered over the variable.

```ts
// https://github.com/Automattic/mongoose/blob/9875dbf53be7b1a4652ab3ac57db2a881999889a/types/index.d.ts#L529
export type LeanDocument<T> = {
  [K in keyof Omit<T, Exclude<keyof Document, '_id' | 'id' | '__v'> | '$isSingleNested'>]: T[K] extends never[] ? never[] : T[K] extends Document[] ? Array<LeanDocument<T[K][number]>> : T[K] extends unknown[] ? T[K][number] extends Document ? Array<LeanDocument<T[K][number]>> : T[K] : T[K] extends Document ? LeanDocument<T[K]> : T[K];
};
```

```ts
// https://github.com/Automattic/mongoose/blob/9875dbf53be7b1a4652ab3ac57db2a881999889a/types/query.d.ts#L422
lean<LeanResultType = ResultType extends unknown[] ?  Array<LeanDocument<Require_id<RawDocType>>> : LeanDocument<Require_id<RawDocType>> | null>(val?: boolean | any): QueryWithHelpers<LeanResultType, DocType, THelpers, RawDocType>;
```

I've tested it with your tests defined in https://github.com/Automattic/mongoose/blob/9875dbf53be7b1a4652ab3ac57db2a881999889a/test/types/lean.test.ts and it works as well.

#### This is what it looks like with the current `LeanDocument` definition

![Screenshot 2022-06-06 at 13 36 04](https://user-images.githubusercontent.com/55806855/172161985-1b9186e4-5e5c-4adf-96d2-1556e9cd9c15.png)

Notice the 52+ properties omitted from the result

---

Also, In the definition for the `Require_id`

https://github.com/Automattic/mongoose/blob/9875dbf53be7b1a4652ab3ac57db2a881999889a/types/index.d.ts#L94

After checking if it extends a `_id`, doing this `(T & { _id: T['_id'] })` again rather than just using the `T` type since it already has the `_id` property only causes it to look more unorganised in the hover tooltip.

#### Using `(T & { _id: T['_id'] })`
![Screenshot 2022-06-06 at 13 17 11](https://user-images.githubusercontent.com/55806855/172159757-ef1a9dcd-566a-4bf1-b27d-5e542e29369a.png)

#### Using `T`
![Screenshot 2022-06-06 at 13 18 57](https://user-images.githubusercontent.com/55806855/172159793-bfbd3edd-9a32-4844-ba1c-500fa4265214.png)

Please let me know what you think about this.

If there is something you'll like to change or a reason why you wouldn't want this change.

Otherwise, if it's fine and you have no other worries about the changes, I can open a PR for it. Thanks

### Steps to Reproduce

1. Make an `ObjectId` type with bson's definition and mongoose's definition
2. Define a Schema using the `ObjectId` above as its `_id` type
3. Query that Model and `lean` the result.
4. Attempt to use the `.equals` method with a string or another variable.

```ts
import { ObjectId } from "bson";
import { Schema } from "mongoose";

type Book = {
  _id: ObjectId & Schema.Types.ObjectId;
  title: string;
}

const BookSchema = new Schema<Book>({ title: { type: String } });
const BookModel = model("book", BookSchema);

async function run() {
  const { _id } = await BookModel.create({
    title: "Book Title 👍"
  });
  const res = await BookModel.findOne({}).lean();
  res?._id.equals(_id) // TS Error
}

run();
```

### Expected Behavior

Not to have the Typescript error.

**Comment by vkarpov15** [comment]
The below script compiles fine with Mongoose 6.3.5 and TypeScript 4.7.3, which version of TS are you using?

```
import { ObjectId } from "bson";
import { Types, Schema, model } from "mongoose";

type Book = {
  _id: ObjectId & Schema.Types.ObjectId;
  title: string;
}

const BookSchema = new Schema<Book>({ title: { type: String } });
const BookModel = model("book", BookSchema);

async function foo() {
  const { _id } = await BookModel.create({
    title: "Book Title 👍"
  });
  const res = await BookModel.findOne({}).lean();
  res?._id.equals(_id) 
}
```

Also, I'd recommend changing your document interface to the following. The `& Schema.Types.ObjectId` part is incorrect, because `Schema.Types.ObjectId` is a schematype, not an ObjectId.

```
import { Types, Schema, model } from "mongoose";

type Book = {
  _id: Types.ObjectId;
  title: string;
}
```

**Comment by iammola** [comment]
Thanks for telling me about the `Schema.Types.ObjectId` schematype.

Yeah I don't know why I gave such a simple repro script. The current types worked for me for top-level ObjectId's. But not nested ones.

I'm not sure If you'll prefer I update the script in my original question, but here is one that I have an error with.

In this one, books are grouped by year in the database and the authors are in a different model.

```ts
import { ObjectId } from "bson";
import { Schema, model } from "mongoose";

type Book = {
  _id: ObjectId & Schema.Types.ObjectId;
  title: string;
  authors: Array<ObjectId & Schema.Types.ObjectId>;
};

type Year = {
  _id: ObjectId & Schema.Types.ObjectId;
  year: number;
  books: Book[];
};

type Author = {
  name: string;
  dob: Date;
};

const BookSchema = new Schema<Book>({
  title: { type: String },
  authors: {
    type: [Schema.Types.ObjectId],
    ref: "author",
  },
});

const YearSchema = new Schema<Year>({
  year: {
    type: Number,
  },
  books: { type: [BookSchema] },
});

const AuthorSchema = new Schema<Author>({
  name: { type: String, required: true },
  dob: { type: Date },
});

const AuthorModel = model("author", AuthorSchema);
const YearModel = model("year", YearSchema);

async function run() {
  const [klopp, wolfgang, arrigo] = await AuthorModel.create([
    {
      name: "Jürgen Klopp",
      dob: new Date(1967, 5, 16),
    },
    {
      name: "Wolfgang Frank",
      dob: new Date(1951, 1, 21),
    },
    {
      name: "Arrigo Sacchi",
      dob: new Date(1946, 3, 1),
    },
  ]);

  const booksIn2022 = await YearModel.create({
    year: 2022,
    books: [
      {
        title: "The Art of the Gengen-Press",
        authors: [klopp._id, wolfgang._id, arrigo._id],
      },
      {
        title: "The Benefits of a High Defence Line?",
        authors: [klopp._id],
      },
    ],
  });

  const result = await YearModel.findOne({ year: 2022 }).lean();
  if (result === null) return;

  result._id.equals(booksIn2022._id) // No TS Error
  result.books[0].authors[0].equals(klopp._id); // TS Error
}

run();
```

There isn't a TS error for the document's ID. It only affects the deeper nested ObjectId properties. But if I use the types I suggested above for the `LeanDocument`, it works fine.

**Comment by vkarpov15** [comment]
@iammola why not just use `Types.ObjectId` as shown below? I can confirm your script compiles with the below changes:

```ts
import { Schema, Types, model } from "mongoose";

type Book = {
  _id: Types.ObjectId;
  title: string;
  authors: Array<Types.ObjectId>;
};

type Year = {
  _id: Types.ObjectId;
  year: number;
  books: Book[];
};

type Author = {
  name: string;
  dob: Date;
};

const BookSchema = new Schema<Book>({
  title: { type: String },
  authors: {
    type: [Schema.Types.ObjectId],
    ref: "author",
  },
});

const YearSchema = new Schema<Year>({
  year: {
    type: Number,
  },
  books: { type: [BookSchema] },
});
```

**Comment by iammola** [comment]
@vkarpov15

After I reposted the script, I tried the `Types.ObjectId` definition and it was exactly what I was trying to get from extending with `bson`'s definitions. I don't know how I forgot to comment on it, so I'm sorry for that.

With that said, I'll appreciate if you still consider improving the visibility of the lean type when the variable is hovered over. It's not as important, but personally it's something that I feel will be massively appreciated. 

Thanks again for your work on the package.
